File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

bool solution(vector<string> phone_book) {
    
    // 1. HashMap 만들기
    unordered_map<string, int> map;
    
    for (int i = 0; i < phone_book.size(); i++)
        map[phone_book[i]] = 1;
    
    // 2. 모든 전화번호의 substring이 HashMap에 존재 유무 파악
    for (int i = 0; i < phone_book.size(); i++)
    {
        for (int j = 0; j < phone_book[i].size() - 1; j++)
        {
            string phone_number = phone_book[i].substr(0, j + 1);
            if (map[phone_number])
                return false;
        }
    }
    
    // 3. 여기까지 오면 접두어 없음
    return true;
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;
import java.io.*;

class Solution {
    public boolean solution(String[] phone_book) {
        HashMap<String, Integer> hs = new HashMap<>();
        Arrays.sort(phone_book);
        
        for (int i = 0; i < phone_book.length; i++) {
            StringBuilder st = new StringBuilder();
            for (int j = 0 ; j < phone_book[i].length(); j++) {
                st.append(phone_book[i].charAt(j));
                if (hs.containsKey(st.toString())) return false;
            }
            String change_st = st.toString();
            hs.put(change_st, 1);
        }
        return true;
    }
}